README.md
# Belgian Weather, UV-Index per day per province by using OpenWeatherMap.org and BPost.be data sources

The software collects the Weather and the UV-Index of the Belgian communes from the OpenWeatherMap database. See [https://www.openweathermap.org](https://www.openweathermap.org)

The weather data will be linked to other data to answer one question: "Does the usage of the Weather or UV-Index data influence (or not) the model forecasting the spread of the COVID-19?"

# How to build the DataFrame by using OpenWeatherMap.org and BPost.be databases?

1. Set the api_key from [https://home.openweathermap.org/api_keys](https://home.openweathermap.org/api_keys)
2. Install python requirements by running `pip install -r requirements.txt`
3. Run `python src/eosc-gees-weather_in_belgian_provinces_per_day.py`
4. Look in the [output_directory/data](output_directory/data) directory to manipulate .csv and .json files
5. Manipulate the aggregated version of the data on [output_directory/data/latest/by_province_and_quartile.csv](output_directory/data/latest/by_province_and_quartile.csv)

# Use crontab to generate the data and commit the new dataset on github

* run  `python ~/dev/github/OpenWeatherMap/build_belgian_weather_by_province.py`

# What are the data available?

* Look into the directory [export_directory/data/by_date/yyyy-mm-dd](export_directory/data/by_date/yyyy-mm-dd)/*.{json|csv}` containing OpenWeatherMap.org for Belgian cities

* Weather data is grouped by province and per day
* We have computed the quantiles 25-50-75 for the following weather conditions: 
    * main.feels_like: See [https://openweather.co.uk/blog/post/new-feels-temperature-openweather-apis](https://openweather.co.uk/blog/post/new-feels-temperature-openweather-apis)
    * main.humidity
    * main.pressure
    * main.temp_max: maximal temperature
    * main.temp_min: minimal temperature
    * main.temp: temperature
    * wind.deg
    * wind.speed
    * uv-index

# The context

Given the consequences of the COVID-19 pandemic for public health, many stakeholders in public and private sectors engage in global efforts to treat and understand those exposed to the virus and to contain the outbreak.

Universities and private companies contribute to the analysis of data useful to forecast the spread of the "Severe Acute Respiratory Syndrome CoronaVirus 2" (SARS-CoV-2).

Those persons include researchers and consultants who share their analysis to the public administration and members of governments who will shape the rules and regulations describing how the citizens should behave during the pandemic.

# Why collecting the Weather and UV-Index data?

In Belgium, the  Group of Experts for an Exit Strategy (GEES) drafts proposals to envisage a gradual deconfinement. The GEES supports the [Belgian Federal Government](https://www.belgium.be/en/about_belgium/government/federal_authorities/federal_government).

See also [https://www.info-coronavirus.be/en](https://www.info-coronavirus.be/en)

The GEES uses data from researchers, amongst others. Those researchers build models of the spread of the virus by using diverse data: mobility, health, beds in hospitals, fatalities, current regulation ...

One question remains: **does the usage of the Weather or UV-Index data influence (or not) the spread of the virus?**

One of those groups of researchers includes the [Machine Learning Group](https://mlg.ulb.ac.be) (MLG) from the [Université Libre de Bruxelles](https://www.ulb.be). The MLG supports the initiative by providing models made of Artificial Intelligence/Machine Learning algorithms. The models built by the MLG influence the Exit Strategy.

# Technical challenges

OpenWeatherMap does not store the Postal code of the communes in Belgium. The Belgian Post provides the Belgian postal codes in French and Dutch [https://www.bpost.be](https://www.bpost.be)

This project joins both databases to enable a link between diverse types of data: mobility, postal codes, etc.

# The tools used to build the software

* Pandas [https://pandas.pydata.org](https://pandas.pydata.org)
* Python [https://www.python.org](https://www.python.org)
* Numpy [https://numpy.org](https://numpy.org)
* Kaggle [https://www.kaggle.com](https://www.kaggle.com)
* Visual Studio Code [https://code.visualstudio.com](https://code.visualstudio.com)

# The sponsors of the project

## The EOSC Secretariat

The **EOSC Secretariat** supports the governance of the **E**uropean **O**pen **S**cience **C**loud (EOSC). See [https://www.eoscsecretariat.eu/](https://www.eoscsecretariat.eu)

The EOSC supports projects aiming to make data **F**indable, **A**ccessible, **I**nteroperable, and **R**eproducible (FAIR) for scientists; these combinations would lead to (unforeseen) reuse and faster development of science. 

EOSC Secretariat granted funds to [http://www.alt-f1.be](http://www.alt-f1.be) to create a "Crossroads Data Bank for COVID-19".

Read the latest news from the European Commission regarding the Open Science Cloud initiative: [https://ec.europa.eu/research/openscience/index.cfm?pg=open-science-cloud](https://ec.europa.eu/research/openscience/index.cfm?pg=open-science-cloud) 

## OpenWeatherMap

**OpenWeatherMap** provides, free-of-charge, access to weather data to its historical weather data [https://openweathermap.org/api](https://openweathermap.org/api)

OpenWeather Ltd is a British-based tech company that provides weather and satellite data worldwide. OpenWeather collects and processes raw data from a variety of sources, and gives its customers access to the archive. See [https://openweathermap.org/](https://openweathermap.org/)

See [Openweather LTD Helps The Fight To Overcome COVID-19](https://bit.ly/2ZohgFF)

# The supporters of the project

## The Machine Learning Group

**The Machine Learning Group**, founded in 2004 by [Gianluca Bontempi](https://mlg.ulb.ac.be/wordpress/members-2/gianluca-bontempi),  is a research unit of the Computer Science Department of the ULB (Université Libre de Bruxelles, Brussels, Belgium), Faculty of Sciences, currently co-headed by Prof. Gianluca Bontempi and Prof. [Tom Lenaerts](https://mlg.ulb.ac.be/wordpress/members-2/tom-lenaerts).

MLG targets machine learning and behavioral intelligence research focusing on time series analysis, big data mining, causal inference, network inference, decision-making models, and behavioral analysis with applications in data science, medicine, molecular biology, cybersecurity and social dynamics related to cooperation, emotions, and others.

See [https://mlg.ulb.ac.be/](https://mlg.ulb.ac.be/)

# Legal matters

The current software digests data that do not contain personal data.

The output of this project is linked to data that do not contain Personally Identifiable Information (PII). See [https://en.wikipedia.org/wiki/Personal_data](https://en.wikipedia.org/wiki/Personal_data)

# The data models

## Belgian Postal Codes

* Postal codes in French [https://www.bpost.be/site/fr/envoyer/adressage/rechercher-un-code-postal](https://www.bpost.be/site/fr/envoyer/adressage/rechercher-un-code-postal/)
* Postal codes in Dutch [https://www.bpost.be/site/nl/verzenden/adressering/zoek-een-postcode](https://www.bpost.be/site/nl/verzenden/adressering/zoek-een-postcode)

Here are the columns describing the postal codes: 

```
Index(['country code', 'postal code', 'place name', 'admin name1',
       'admin code1', 'admin name2', 'admin code2', 'admin name3',
       'admin code3', 'rounded_lat', 'rounded_lon', 'accuracy'],
      dtype='object')
```

## OpenWeatherMap List of cities IDs

Here are the columns describing the cities for which OpenWeatherMap tracks the Weather: [http://bulk.openweathermap.org/sample/history.city.list.json.gz](http://bulk.openweathermap.org/sample/history.city.list.json.gz)

```
Index(['id', 'city.id.$numberLong', 'city.name', 'city.findname',
       'city.country', 'city.coord.lon', 'city.coord.lat',
       'city.zoom.$numberLong', 'id.$numberLong', 'city.coord.lon.$numberLong',
       'city.coord.lat.$numberLong', 'rounded_lat', 'rounded_lon'],
      dtype='object')
```

## OpenWeatherMap Historical data

OpenWeatherMap provides an API returning the hourly Weather. A call to the API returns a maximum of 7 consecutive days of data.
[https://openweathermap.org/history](https://openweathermap.org/history) 

Here is an example of a call: `http://history.openweathermap.org/data/2.5/history/city?id={id}&type=hour&start={start}&end={end}&appid={YOUR_API_KEY}`

### The response from the history of OpenWeatherMap

The documentation of the **response from the history** of OpenWeatherMap is here: [https://openweathermap.org/weather-data](https://openweathermap.org/weather-data)

``` json
{
    "dt": 1587312000,
    "main": {
        "temp": 287.44,
        "feels_like": 284.3,
        "pressure": 1015,
        "humidity": 76,
        "temp_min": 287.04,
        "temp_max": 288.15
    },
    "wind": {
        "speed": 4.6,
        "deg": 40
    },
    "clouds": {
        "all": 77
    },
    "weather": [
        {
            "id": 500,

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.4.72](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.71...v3.4.72) (2020-10-17)


### Documentations

* update with Weather and UV-Index for October 16, 2020 ([9331ad8](https://github.com/ALT-F1/OpenWeatherMap/commit/9331ad8e954bc584f10bd95df5113279725284d7))

### [3.4.71](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.70...v3.4.71) (2020-10-16)


### Documentations

* update with Weather and UV-Index for October 15, 2020 ([f0509ca](https://github.com/ALT-F1/OpenWeatherMap/commit/f0509ca25966d4400d7fa82698553778d5e3c134))

### [3.4.70](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.69...v3.4.70) (2020-10-15)


### Documentations

* update with Weather and UV-Index for October 14, 2020 ([dcfa636](https://github.com/ALT-F1/OpenWeatherMap/commit/dcfa636bc3ae0424b0d69ca9e898fcb2dbd9920c))

### [3.4.69](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.68...v3.4.69) (2020-10-14)


### Documentations

* update with Weather and UV-Index for October 13, 2020 ([0bae8fa](https://github.com/ALT-F1/OpenWeatherMap/commit/0bae8facfd8f9db1b10a24aae63ba5f3552693fe))

### [3.4.68](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.67...v3.4.68) (2020-10-13)


### Documentations

* update with Weather and UV-Index for October 12, 2020 ([faf24f2](https://github.com/ALT-F1/OpenWeatherMap/commit/faf24f2ca09fb1d8356dd128a3fd947acf4fc469))

### [3.4.67](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.66...v3.4.67) (2020-10-12)


### Documentations

* update with Weather and UV-Index for October 11, 2020 ([16d861e](https://github.com/ALT-F1/OpenWeatherMap/commit/16d861ecb705b65eb9fbdbcf28b6692f020ee723))

### [3.4.66](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.65...v3.4.66) (2020-10-11)


### Documentations

* update with Weather and UV-Index for October 10, 2020 ([4413542](https://github.com/ALT-F1/OpenWeatherMap/commit/4413542ee7426344bc8a95cb8cb541da7b9405b6))

### [3.4.65](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.64...v3.4.65) (2020-10-10)


### Documentations

* update with Weather and UV-Index for October 09, 2020 ([667af4c](https://github.com/ALT-F1/OpenWeatherMap/commit/667af4cbbb30c2344f91f9326153bdfb06f7c127))

### [3.4.64](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.63...v3.4.64) (2020-10-09)


### Documentations

* update with Weather and UV-Index for October 08, 2020 ([a2a58d9](https://github.com/ALT-F1/OpenWeatherMap/commit/a2a58d91a0bf5e3f727530b1da571b695ab40c6d))

### [3.4.63](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.62...v3.4.63) (2020-10-08)


### Documentations

* update with Weather and UV-Index for October 07, 2020 ([0338259](https://github.com/ALT-F1/OpenWeatherMap/commit/03382596a829a0e4e983d28ab3b33b8d167c4676))

### [3.4.62](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.61...v3.4.62) (2020-10-07)


### Documentations

* update with Weather and UV-Index for October 06, 2020 ([8d4c755](https://github.com/ALT-F1/OpenWeatherMap/commit/8d4c7551f08456078cd8beeb36f9f57b45e07e9f))

### [3.4.61](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.60...v3.4.61) (2020-10-06)


### Documentations

* update with Weather and UV-Index for October 05, 2020 ([38aaf32](https://github.com/ALT-F1/OpenWeatherMap/commit/38aaf32717d7ae7bc54bfe5b47675cca028bfb98))

### [3.4.60](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.59...v3.4.60) (2020-10-05)


### Documentations

* update with Weather and UV-Index for October 04, 2020 ([9c1baae](https://github.com/ALT-F1/OpenWeatherMap/commit/9c1baaee1ea5402e4a9e226994dda04d88993e6d))

### [3.4.59](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.58...v3.4.59) (2020-10-04)


### Documentations

* update with Weather and UV-Index for October 03, 2020 ([9f7bb6a](https://github.com/ALT-F1/OpenWeatherMap/commit/9f7bb6a933981a90875fcb268de639a165fd1f71))

### [3.4.58](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.57...v3.4.58) (2020-10-02)


### Documentations

* update with Weather and UV-Index for October 01, 2020 ([59c0796](https://github.com/ALT-F1/OpenWeatherMap/commit/59c07967234aa5ddb189b2724a73b377ecee4a33))

### [3.4.57](https://github.com/ALT-F1/OpenWeatherMap/compare/v3.4.56...v3.4.57) (2020-10-01)


### Documentations


src/openweathermap_helpers.py
# %% [code]
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

# initialize libraries
import logging
import http.client
from datetime import datetime, timedelta, date
import json
from altf1be_helpers import AltF1BeHelpers
from pytz import timezone
from os import path
import numpy as np  # linear algebra
import pandas as pd  # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory
import sys
import os
import socket
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All"
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# %% [code]

filename = os.path.join(AltF1BeHelpers.output_directory(
    ['logs']), 'openweathermap_helpers.py.log')

logging.basicConfig(filename=filename, level=logging.DEBUG)
logging.info(f"Log file is stored here : {filename}")

# initialize constants
DATA_FROM_FUTURE_IS_UNAVAILABLE = 400000


class OpenWeatherMap():
    """The class provides a read-to-use OpenWeatherMap functionalities

    """

    def get_historical_uv_http(self, lat, lon, cnt, start_date, end_date):
        """Download the UV-Index (Ultraviolet) from OpenWeatherMap.org

        See: https://en.wikipedia.org/wiki/Ultraviolet_index

        Sample url: https://api.openweathermap.org/data/2.5/uvi/history?appid={{your_api_key}}&lat={{lat}}&lon={{lon}}&cnt={{cnt}}&start={{start date}}&end={{end date}}

        Returns
        -------

        [
            {
                "lat": 50.850449,
                "lon": 4.34878,
                "date_iso": "2020-04-24T12:00:00Z",
                "date": 1587729600,
                "value": 5.55
            }
        ]

        """

        conn = http.client.HTTPSConnection("api.openweathermap.org")
        payload = ''
        headers = {}
        url = f"/data/2.5/uvi/history?appid={self.secret_api_key}&lat={lat}&lon={lon}&cnt={cnt}&start={start_date}&end={end_date}"

        try:
            conn.request(
                "GET",
                url,
                payload,
                headers
            )
        except socket.gaierror:
            msg1 = f"{AltF1BeHelpers.hide_secrets_from_url(url)}"
            msg2 = f"get_historical_uv: socket.gaierror error: {sys.exc_info()}"
            print(f"{msg1}\n{msg2}")
            logging.error(msg1)
            logging.error(msg2)
        except:
            msg1 = f"{AltF1BeHelpers.hide_secrets_from_url(url)}"
            msg2 = f"get_historical_uv: {sys.exc_info()}"
            print(f"{msg1}\n{msg2}")
            logging.error(msg1)
            logging.error(msg2)

        res = conn.getresponse()
        data = res.read()
        result = data.decode("utf-8")
        # print(result)
        return result

    def get_historical_uv(self, lat, lon, cnt, start_date, end_date):
        """Download the UV-Index (Ultraviolet) from OpenWeatherMap.org

        See: https://en.wikipedia.org/wiki/Ultraviolet_index

        Sample url: https://api.openweathermap.org/data/2.5/uvi/history?appid={{your_api_key}}&lat={{lat}}&lon={{lon}}&cnt={{cnt}}&start={{start date}}&end={{end date}}

        Returns
        -------

        [
            {
                "lat": 50.850449,
                "lon": 4.34878,
                "date_iso": "2020-04-24T12:00:00Z",
                "date": 1587729600,
                "value": 5.55
            }
        ]

        """
        protocol = "https://"
        base_url = "api.openweathermap.org"
        url = f"/data/2.5/uvi/history?appid={self.secret_api_key}&lat={lat}&lon={lon}&cnt={cnt}&start={start_date}&end={end_date}"
        url_to_call = f"{protocol}{base_url}{url}"
        res = AltF1BeHelpers.requests_retry_session().get(url_to_call)
        logging.info(
            f"http status: {res.status_code} for {AltF1BeHelpers.hide_secrets_from_url(url)}")

        data = res.content
        result = data.decode("utf-8")
        # print(result)
        return result

    def save_uv_index_to_file(self, current_row, year=2020, month=5, day=1, format='csv'):
        """
            get the UV Index for a specific day from OpenWeatherMap.org
        """

        city_name = current_row['city.findname']
        city_id = current_row['id']

        # start_datetime = datetime.datetime(
        #     year=year, month=month, day=day, hour=0, minute=0
        # )
        start_datetime = datetime(
            year=year, month=month, day=day, hour=0, minute=0
        ).replace(
            tzinfo=timezone('UTC')
        )

        end_datetime = start_datetime

        print(
            f'save_uv_index_to_file : save file: {start_datetime.strftime("%Y-%m-%d")}')

        filename = os.path.join(
            self.altF1BeHelpers.output_directory(
                [
                    'OpenWeatherMap.org',
                    'uv-index',
                    start_datetime.strftime("%Y-%m-%d"),
                ]
            ),
            f'{city_name}-{start_datetime.strftime("%Y-%m-%d")}-uv_index'
        )

        if os.path.exists(f"{filename}.json"):
            logging.warning(
                f"WARNING: UV-Index file already exists: We skip its retrieval from OpenWeathMap.org: {filename}.json"
            )
            with open(f"{filename}.json", "r") as json_file:
                uv_index = json.load(json_file)
            return uv_index

        # logging.info(f"UV-Index file stored in : {os.path.dirname(filename)}")

        # Get the UV-Index from OpenWeatherMap.org
        uv_index = self.get_historical_uv(
            lat=current_row['city.coord.lat'],
            lon=current_row['city.coord.lon'],
            cnt=1,
            start_date=int(start_datetime.timestamp()),
            end_date=int(end_datetime.timestamp())
        )

        df_csv = pd.DataFrame()
        if format == 'csv':
            df_csv = pd.concat(
                [df_csv, self.uv_index_json_str_to_flat_df(uv_index)])

            df_csv.to_csv(f"{filename}.csv", index=False)
            logging.info(f"File stored in : {filename}.csv")

        with open(f"{filename}.json", 'w') as outfile:
            json.dump(uv_index, outfile, indent=2)
            logging.info(f"File stored in : {filename}.json")

        return uv_index

    def get_historical_weather_http(self, city_id, start, end):
        """Download the weather data from OpenWeatherMap.org

        See https://openweathermap.org/history

        Parameters
        ----------
        city_id : int
            The city id recognized by OpenWeatherMap.org
        start : start date (unix time, UTC time zone), e.g. start=1369728000
        end : end date (unix time, UTC time zone), e.g. end=1369789200

        Returns
        -------
        json string
            json string containing the weather for a specific date

        """
        conn = http.client.HTTPSConnection("history.openweathermap.org")
        payload = ''
        headers = {}
        conn.request(
            "GET", f"/data/2.5/history/city?id={city_id}&start={start}&end={end}&appid={self.secret_api_key}", payload, headers)
        res = conn.getresponse()
        data = res.read()
        result = data.decode("utf-8")
        # print(result)
        return result

    def get_historical_weather(self, city_id, start, end):
        """Download the weather data from OpenWeatherMap.org

        See https://openweathermap.org/history

        Parameters
        ----------
        city_id : int
            The city id recognized by OpenWeatherMap.org
        start : start date (unix time, UTC time zone), e.g. start=1369728000
        end : end date (unix time, UTC time zone), e.g. end=1369789200

        Returns
        -------
        json string
            json string containing the weather for a specific date

        """

        protocol = "https://"
        base_url = "history.openweathermap.org"
        url = f"/data/2.5/history/city?id={city_id}&start={start}&end={end}&appid={self.secret_api_key}"
        url_to_call = f"{protocol}{base_url}{url}"

        res = AltF1BeHelpers.requests_retry_session().get(url_to_call)

        logging.info(
            f"http status: {res.status_code} for {AltF1BeHelpers.hide_secrets_from_url(url)}"
        )

        data = res.content
        result = data.decode("utf-8")
        # print(result)
        return result

    def save_weather_to_file(self, current_row, year=2020, month=5, day=1, format='csv'):
        """Get the weather for a specific day from OpenWeatherMap.org
        """

        city_name = current_row['city.findname']
        city_id = current_row['id']

        start_datetime, end_datetime = self.get_range_between_days(
            year, month, day)

        filename = os.path.join(
            self.altF1BeHelpers.output_directory(
                [
                    'OpenWeatherMap.org',
                    start_datetime.strftime("%Y-%m-%d")
                ]
            ),
            f'{city_name}-{start_datetime.strftime("%Y-%m-%d")}-weather'
        )

        if os.path.exists(f"{filename}.json"):
            logging.warning(
                f"WARNING: Weather file already exists: We skip its retrieval from OpenWeathMap.org: {filename}.json")
            with open(f"{filename}.json", "r") as json_file:
                weather_json = json.load(json_file)
            return weather_json

        # logging.info(f"Weather file stored in : {os.path.dirname(filename)}")

        # get the weather from OpenWeatherMap.org
        weather_json = self.get_historical_weather(
            city_id, int(start_datetime.timestamp()), int(end_datetime.timestamp()))

        df_csv = pd.DataFrame()
        if format == 'csv':
            df_csv = pd.concat(
                [df_csv, self.weather_json_str_to_flat_df(weather_json)])

            df_csv.to_csv(f"{filename}.csv", index=False)
            logging.info(f"Weather file stored in : {filename}.csv")

        with open(f"{filename}.json", 'w') as outfile:
            json.dump(weather_json, outfile, indent=2)
            logging.info(f"Weather file stored in : {filename}.json")

        return weather_json

    def weather_csv_to_df(self, json_data):
        """Convert a CSV into a Dataframe

        """
        # {'code': 400000, 'message': 'data from future is...available'}
        df_csv = pd.DataFrame(columns=['message',
                                       'cod',
                                       'city_id',
                                       'calctime',
                                       'cnt',
                                       'dt',
                                       'main.temp',
                                       'main.feels_like',
                                       'main.pressure',
                                       'main.humidity',
                                       'main.temp_min',
                                       'main.temp_max',
                                       'wind.speed',
                                       'wind.deg',
                                       'clouds.all',
                                       'weather.id',
                                       'weather.main',
                                       'weather.description',
                                       'weather.icon', ])

        try:
            if json_data["code"] == DATA_FROM_FUTURE_IS_UNAVAILABLE:
                return df_csv
        except KeyError:
            # openweathermap sent a reponse including wheater data
            pass
        else:
            logging.error("TODO: handle this error")
            logging.erro("json_data: {json_data}")
            logging.error("Unexpected error:", sys.exc_info()[0])
            raise

        message = json_data["message"]
        cod = json_data["cod"]
        city_id = json_data['city_id']
        calctime = json_data["calctime"]
        cnt = json_data["cnt"]

        for p in json_data['list']:

            for weather in p["weather"]:

                df_new_row = pd.DataFrame({
                    'message': [message],
                    'cod': [cod],
                    'city_id': [city_id],
                    'calctime': [calctime],
                    'cnt': [cnt],

                    'dt': p["dt"],
                    'main.temp': p["main"]["temp"],
                    'main.feels_like': p["main"]["feels_like"],
                    'main.pressure': p["main"]["pressure"],
                    'main.humidity': p["main"]["humidity"],
                    'main.temp_min': p["main"]["temp_min"],
                    'main.temp_max': p["main"]["temp_max"],

                    'wind.speed': p["wind"]["speed"],
                    'wind.deg': p["wind"]["deg"],

                    'clouds.all': p["clouds"]["all"],

                    'weather.id': weather["id"],
                    'weather.main': weather["main"],
                    'weather.description': weather["description"],
                    'weather.icon': weather["icon"],
                })

            df_csv = df_csv.append(df_new_row, ignore_index=True)

        return df_csv

    def weather_json_str_to_flat_df(self, json_str):
        """Convert a JSON string into a flat DataFrame

        """
        j = json.loads(json_str)
        df_csv = self.weather_csv_to_df(j)

        df_csv['dt_str'] = pd.to_datetime(df_csv['dt'], unit='s')
        return df_csv

    def uv_index_json_str_to_flat_df(self, json_str):
        """Convert a UV-Index JSON string into a flat DataFrame

        """

        # ACTION: to remove if the assignation to df makes no difference : df_csv = df = pd.read_json(json_str, convert_dates=False)
        df_csv = pd.read_json(json_str, convert_dates=False)
        return df_csv

    def get_range_between_days(self, year, month, day):
        """Get a range of dates between 2 dates

        Returns
        -------
        unix time, UTC time zone
            start_datetime e.g. end=1369789200
            end_datetime e.g. end=1369789200

        """

        start_datetime = datetime(
            year=year, month=month, day=day, hour=0, minute=0
        ).replace(
            tzinfo=timezone('UTC')
        )
        printable_start_datetime = start_datetime.strftime("%Y-%m-%d_%Hh%M")

        end_datetime = datetime(
            year=year, month=month, day=day, hour=23, minute=0
        ).replace(
            tzinfo=timezone('UTC')
        )
        printable_end_datetime = end_datetime.strftime("%Y-%m-%d_%Hh%M")

        # logging.info(
        #     f"get_range_between_days: {printable_start_datetime}, {printable_end_datetime}"
        # )

        return start_datetime, end_datetime

    def get_range_in_a_month(self, year, month, current_week, days_in_a_week, days_in_current_month):
        """
            get a range in a month 
        """
        start_datetime = datetime.datetime(
            year=year, month=month, day=1+(current_week*days_in_a_week), hour=0, minute=0
        ).replace(
            tzinfo=timezone('UTC')
        )
        printable_start_datetime = start_datetime.strftime("%Y-%m-%d_%Hh%M")

        end_day = days_in_current_month if 7 + \
            (current_week*days_in_a_week) > days_in_current_month else 7+(current_week*days_in_a_week)

        end_datetime = datetime.datetime(
            year=year, month=month, day=end_day, hour=23, minute=0
        ).replace(
            tzinfo=timezone('UTC')
        )
        printable_end_datetime = end_datetime.strftime("%Y-%m-%d_%Hh%M")

        logging.info(
            f"get_range_in_a_month: {printable_start_datetime}, {printable_end_datetime}"
        )

        return start_datetime, end_datetime

    def save(self):

        self.df_cities_weather_in_be.to_excel(os.path.join(
            self.altF1BeHelpers.output_directory(['OpenWeatherMap.org']), "df_cities_weather_in_be.xlsx"))

    def get_openweathermap_secret_key(self):
        self.secret_api_key = None
        # Set the api_key and the paths on a Kaggle
        try:
            from kaggle_secrets import UserSecretsClient
            user_secrets = UserSecretsClient()
            self.secret_api_key = user_secrets.get_secret("API_KEY")
            print(f"OpenWeatherMap.org secret key found inside the Kaggle secret keys")
        except KeyError:
            print("KeyError: the secret key does not exist on Kaggle")
            # print("(Un)expected error:", sys.exc_info()[0])
        except ModuleNotFoundError:
            print("INFO: you may not be running kaggle, the OpenWeaterMap.org API KEY will be collected from the environment variable")
            # print("(Un)expected error:", sys.exc_info()[0])

        # Set the api_key and the paths on a regular OS
        if (self.secret_api_key == None):
            try:
                self.secret_api_key = os.environ['OPENWEATHERMAP_API_KEY']
                print(
                    f"OpenWeatherMap.org secret key found inside the environment variable")
            except KeyError:
                print("INFO: OPENWEATHERMAP_API_KEY is not a environment variable")
                print("You may use Kaggle to run this code")
                print(
                    "Consider adding an environment variable if OpenWeatherMap.org triggers an error related to the API TOKEN")

        if (self.secret_api_key == None):
            logging.error(
                f"ERROR: OpenWeatherMap.org secret was NOT found inside Kaggle nor the environment variables")

        return self.secret_api_key

    def get_openweathermap_paths(self):
        """
        initialize three paths history_city_list_path, postal_codes_in_be_from_geonames_org
        """
        history_city_list_path = "kaggle/input/historycitylistjson/history.city.list.json"
        if (self.altF1BeHelpers.is_interactive()):
            # source https://openweathermap.org/history
            self.history_city_list_path = f"/{history_city_list_path}"
        else:
            self.history_city_list_path = os.path.join(
                os.path.abspath(os.getcwd()),
                "src",
                history_city_list_path
            )

        logging.info(f"history_city_list_path: {self.history_city_list_path}")

        return self.history_city_list_path

    def clean_columns(self, df):
        """

        """

        # set the
        for i, row in df[df['city.coord.lon'].isna()].iterrows():
            df.at[i, 'city.coord.lon'] = row['city.coord.lon.$numberLong']
            # df.loc[index]['city.coord.lon'] = df.loc[index]['city.coord.lon.$numberLong']

        for i, row in df[df['city.coord.lat'].isna()].iterrows():
            row['city.coord.lat'] = row['city.coord.lat.$numberLong']

        # Transform: remove unnecessary columns and rename the column (city.id.$numberLong'->id)
        df = df.drop(columns=[
            'id',
            'city.zoom.$numberLong',
            'city.coord.lon.$numberLong',
            'city.coord.lat.$numberLong',
            'id.$numberLong'
        ])

        # rename "city.id.$numberLong" in id
        df = df.rename(
            columns={'city.id.$numberLong': 'id'})

        return df

    def rename_cities(self, df):
        """
            Make changes in the name of the cities to match the data inside the bpost.be database
        """
        # Transform : Roeulx (OpenWeatherMap) -> Le Roeulx (bpost.be)
        df[df['city.name'] == 'roeulx'] = df[df['city.name']
                                             == 'roeulx'].replace(['roeulx'], ['le roeulx'])

        return df

    def columns_in_lowercase(self, df):
        """
        Transform: column in lowercase
        """
        df['city.findname'] = df['city.findname'].str.lower()
        df['city.name'] = df['city.name'].str.lower()
        return df

    def keep_belgian_cities(self, df):
        """
            Transform: keep the Belgian cities only from the OpenWeatherMap' list
        """
        # filter the belgian cities
        df_cities_weather_in_be = df[df['city.country'] == 'BE'].copy(
            deep=True)
        return df_cities_weather_in_be

    def extract_ww_cities(self):
        """
            extract cities from list of cities recognized by OpenWeatherMap.org
        """
        # load the history city list
        with open(self.history_city_list_path, "r") as f:
            d = json.load(f)

        return pd.json_normalize(d)

    def build_df(self):
        """
            Build the DataFrame containing OpenWeatherMap ready to be manipulated
        """
        df_weather_in_cities = self.extract_ww_cities()
        self.df_cities_weather_in_be = self.keep_belgian_cities(
            df_weather_in_cities)
        self.df_cities_weather_in_be = self.columns_in_lowercase(
            self.df_cities_weather_in_be)
        self.df_cities_weather_in_be = self.rename_cities(
            self.df_cities_weather_in_be)
        self.df_cities_weather_in_be = self.clean_columns(
            self.df_cities_weather_in_be)
        self.save()

    def __init__(self):
        self.altF1BeHelpers = AltF1BeHelpers()
        self.get_openweathermap_secret_key()
        self.get_openweathermap_paths()


# %% [code]
if __name__ == "__main__":

    # instanciate OpenWeatherMap
    openWeatherMap = OpenWeatherMap()
    # print(
    #     f"OpenWeatherMap.org secret_api_key: {(openWeatherMap.secret_api_key)}")
    openWeatherMap.build_df()
    print(openWeatherMap.df_cities_weather_in_be)

    # store the weather of all cities recoginzed by OpenWeatherMap.org in csv and a json formats
    # openWeatherMap.df_cities_weather_in_be.apply(
    #     openWeatherMap.save_weather_to_file, axis=1, args=[2020, 5, 1, 'csv'])

    # store the UV Index of all cities recoginzed by OpenWeatherMap.org in csv and a json formats
    openWeatherMap.df_cities_weather_in_be.apply(
        openWeatherMap.save_uv_index_to_file, axis=1, args=[2020, 3, 1, 'csv'])
